base/math.jl

```diff
@@ -349,8 +349,45 @@ julia> ldexp(5., 2)
 20.0
 ```
 """
-ldexp(x::Float64,e::Integer) = ccall((:scalbn,libm),  Float64, (Float64,Int32), x, Int32(e))
-ldexp(x::Float32,e::Integer) = ccall((:scalbnf,libm), Float32, (Float32,Int32), x, Int32(e))
+function ldexp{T<:AbstractFloat}(x::T, q::Integer)
+    n = Int(q)
+    xu = reinterpret(Unsigned, x)
+    xs = xu & ~sign_mask(T)
+    xs >= exponent_mask(T) && return x # NaN or Inf
+    k = Int(xs >> significand_bits(T))
+    if k == 0 # x is subnormal
+        xs == 0 && return x # +-0
+        m = unsigned(leading_zeros(xs) - exponent_bits(T))
+        xs <<= m
+        xu = xs | (xu & sign_mask(T))
+        k = 1 - signed(m)
+        if n < -50000 # underflow
+            # otherwise may have have an integer underflow in the following k += n computation
+            return flipsign(T(0.0), x)
+        end
+    end
+    k += n
```

Comment:
This would add another branch, is it worth it, because you still have to check that `k` is not overflowing and underflowing in addition to this, which currently is cleverly handled.  This adds, based on benchmarking the various branches,  approximately a  1ns penalty. 

Currently we do a lot worse and cast to Int32.  

I don't necessarily have a problem with this, but question if this is the best way to do it. The quest for the most generic function is satisfying, but it  typically entails  performance penalties. 


Internally for my codes  that need this function, I don't use this   since it's a bit too slow in practice but instead use the following:
https://github.com/JuliaMath/Amal.jl/blob/master/src/ldexp.jl
it has some important cases where it doesn't pass all cases, but that's not important since those cases are handeled externally. 




Thoughts?